### PR TITLE
SY-4113: Refactor Arc C++ and Arc Go to Use telem.MonoClock

### PR DESCRIPTION
**Pull request by emilbon99**
# Issue Pull Request

## Linear Issue

[SY-4113](https://linear.app/synnax/issue/SY-4113)

## Description

Replace all bare `telem.Now()` (Go) and `TimeStamp::now()` (C++) calls with `telem.MonoClock` across both Arc runtimes. `MonoClock` guarantees strictly increasing timestamps by bumping +1ns when the wall clock returns the same value.

This fixes two deterministic test failures on Windows in `arc/go/sequence_test.go` (Re-entry specs) caused by Windows' ~15.6ms wall-clock resolution. When a constant node fires twice in quick succession (first activation and re-entry), `telem.Now()` can return identical timestamps, causing `RefreshInputs()` to skip the update.

**Go changes:**
- `arc/go/stl/constant/constant.go` — root cause fix; constant node now uses MonoClock
- `arc/go/stl/channel/state.go` — indexed channel writes use MonoClock
- `arc/go/stl/time/time.go` — WASM `time.now` host function uses MonoClock
- `arc/go/stl/wasm/node.go` — WASM node output timestamps use MonoClock

**C++ changes:**
- `x/cpp/telem/telem.h` — new `MonoClock` class mirroring the Go implementation
- `arc/cpp/stl/constant/constant.h` — constant node uses MonoClock
- `arc/cpp/stl/channel/channel.h` — channel On/Write/Module nodes use MonoClock
- `arc/cpp/stl/channel/state.h` — channel state writes use MonoClock
- `arc/cpp/stl/stable/stable.h` — stable_for node uses MonoClock (replaces custom NowFn)
- `arc/cpp/stl/time/time.h` — time module WASM binding uses MonoClock
- `arc/cpp/runtime/wasm/node.h` — WASM node uses MonoClock

## Basic Readiness

- [ ] I have performed a self-review of my code.
- [ ] I have added relevant, automated tests to cover the changes.
- [ ] I have updated documentation to reflect the changes.

<h3>Greptile Summary</h3>

This PR refactors all bare `telem.Now()` (Go) and `TimeStamp::now()` (C++) calls in the Arc runtime nodes to use per-instance `MonoClock` objects, which guarantee strictly increasing timestamps by bumping +1 ns on duplicate wall-clock readings. The root cause — Windows' ~15.6 ms clock resolution causing identical timestamps on rapid re-entry — is correctly addressed, and both Go and C++ implementations include a tested nil/null source guard.

<h3>Confidence Score: 5/5</h3>

Safe to merge — the change is well-scoped, both language implementations are tested, and previously flagged issues have been addressed.

Only one P2 finding (overly weak assertion in notify_test.cpp), no P0/P1 issues. Core MonoClock logic is correct in both Go and C++, null-source guard is present in C++, and the StableFor nullptr default is now safe.

x/cpp/telem/mono_clock.h — missing deleted copy constructor/assignment (noted in prior review, still unaddressed); x/cpp/notify/notify_test.cpp — weak liveness assertion.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| x/go/telem/mono_clock.go | New file: MonoClock implementation extracted from time_stamp.go, providing monotonically increasing timestamps with nil-source guard. |
| x/cpp/telem/mono_clock.h | New C++ MonoClock header with nullptr guard in constructor (mirrors Go nil-source check); correctly placed at end of telem.h via include; missing deleted copy operations (flagged in prior review thread). |
| arc/cpp/stl/stable/stable.h | StableFor refactored to use MonoClock, replacing custom NowFn; nullptr default parameter is now safe because MonoClock's constructor handles null source with a runtime guard. |
| x/cpp/telem/telem.h | Adds full int64_t arithmetic/comparison operators to TimeStamp; NowFunc typedef moved to mono_clock.h and re-exported via include at end of file — backward compatible. |
| arc/go/stl/constant/constant.go | Constant node now uses per-instance MonoClock instead of bare telem.Now(), fixing duplicate-timestamp issues on Windows. |
| arc/cpp/stl/channel/channel.h | On, Write, and Module nodes each get their own MonoClock instance; write_str lambda correctly captures this to reach Module::clock. |
| x/cpp/notify/notify_test.cpp | ProducerConsumerPattern test fixed to avoid deadlock on signal coalescing; uses done-flag termination and atomic; final assertion EXPECT_GE(received, 1) is very weak but intentional. |
| x/go/telem/mono_clock_test.go | MonoClock tests moved to dedicated file with full coverage of strictly-increasing source, duplicate bumping, backward-going source, and nil-source default. |

</details>

<h3>Flowchart</h3>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A["Arc Node next()"] --> B{Has input timestamps?}
    B -- Yes --> C[Use longest input timestamp]
    B -- No --> D["MonoClock::now() / MonoClock.Now()"]
    D --> E{source_ts <= last?}
    E -- Yes --> F["ts = last + 1ns"]
    E -- No --> G["ts = source_ts"]
    F --> H[last = ts]
    G --> H
    H --> I[Return strictly increasing TimeStamp]
    C --> J[Write output series]
    I --> J
```

<sub>Reviews (5): Last reviewed commit: ["split mono clock into separate files"](https://github.com/synnaxlabs/synnax/commit/d2959fd22870b16338edc62def5e63327bbadab0) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=29721688)</sub>

### CODE REVIEW

File: x/cpp/telem/telem.h [hunk 1]
```diff
@@ -823,6 +892,36 @@ inline void to_proto(const SampleValue &sv, google::protobuf::Value *v) {
 
 using NowFunc = std::function<TimeStamp()>;
 
+/// @brief provides monotonically increasing timestamps. On platforms with
+/// coarse clock resolution (e.g. Windows), consecutive now() calls can return
+/// the same value. MonoClock guarantees each call returns a strictly greater
+/// timestamp than the previous one by bumping by 1 nanosecond when necessary.
+class MonoClock {
```
Comment (pjdotson) [inline x/cpp/telem/telem.h]: should probably be in a separate file instead of making this file even longer

**Review by pjdotson — APPROVED** [review]
